marshmallow_dataclass/__init__.py

```diff
@@ -534,6 +567,39 @@ def new_type(x: _U):
     return new_type
 
 
+# extends Schema for instance check by polyfield
+class SchemaPolyfieldProxy(marshmallow.Schema):
+    """ Proxy class that implement Schema interface to proxify call to a
+    dataclass. It is used in order to disambiguate Union subtype.
+    By convention, we assume the dataclass has one field called "field"
+    """
+    def __init__(self, dataclass):
+        self.schema = dataclass.Schema()
+        self.dataclass = dataclass
+
+    @property
+    def context(self):
+        return self.schema.context
+
+    def dump(self, value):
+        return self.schema.dump(self.dataclass(value))["field"]
+
+    def load(self, value):
+        return self.schema.load({"field": value}).field
+
+    def check_deserialization(self, obj_dict):
+        load = self.schema.load({"field": obj_dict})
+        dump = self.schema.dump(load)["field"]
+        if type(dump) != type(obj_dict):
+            raise TypeError("types do not match")
```

Comment:
can you make this error more explicit? same bellow

Reply:
I've had the 2 types in the string